File: aat-android/src/main/java/ch/bailu/aat/services/elevation/tile/MultiCell4NW.java
package ch.bailu.aat.services.elevation.tile;

public final class MultiCell4NW extends MultiCell {
    /*
     *      a  b
     *      c  D
     */

    private short a, b, c, d;
    private int dzx, dzy;

    private final DemProvider demtile;
    private final int dim;
    private final int total_cellsize;


    public MultiCell4NW(final DemProvider dem) {
        demtile = dem;
        dim = dem.getDim().DIM;
        total_cellsize=Math.round(dem.getCellsize()*4f);
    }

    @Override
    public void set(int x) {
            _set(x);
        dzx=_delta_zx();
        dzy=_delta_zy();
    }


    private void _set(int x) {
        final int b=x-dim;
        final int a=b-1;
        final int c=x-1;
        final int d=x;
        this.a=demtile.getElevation(a);
        this.b=demtile.getElevation(b);
        this.c=demtile.getElevation(c);
        this.d=demtile.getElevation(d);

    }


    @Override
    public int delta_zx() {
        return dzx;
    }

    @Override
    public int delta_zy() {
        return dzy;
    }

    private int _delta_zx() {
        final int sum = ((b + d) - (a + c));
        return  (sum * 100) / total_cellsize;
    }

    private int _delta_zy() {
        final int sum = ((c + d) - (b + a));
        return (sum * 100)  / total_cellsize;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/elevation/tile/MultiCell4NW.kt
package ch.bailu.aat_lib.service.elevation.tile

class MultiCell4NW(private val demProvider: DemProvider) : MultiCell() {
    /*
     *      a  b
     *      c  D
     */
    private var a: Short = 0
    private var b: Short = 0
    private var c: Short = 0
    private var d: Short = 0
    private var dzx = 0
    private var dzy = 0

    private val dim = demProvider.dim.DIM
    private val totalCellSize = Math.round(demProvider.cellsize * 4f)

    override fun set(x: Int) {
        setCell(x)
        setDeltaZX()
        setDeltaZY()
    }

    private fun setCell(x: Int) {
        val b = x - dim
        val a = b - 1
        val c = x - 1
        val d = x
        this.a = demProvider.getElevation(a)
        this.b = demProvider.getElevation(b)
        this.c = demProvider.getElevation(c)
        this.d = demProvider.getElevation(d)
    }

    override fun deltaZX(): Int {
        return dzx
    }

    override fun deltaZY(): Int {
        return dzy
    }

    private fun setDeltaZX() {
        val sum = ((b + d) - (a + c))
        dzx = (sum * 100) / totalCellSize
    }

    private fun setDeltaZY() {
        val sum = ((c + d) - (b + a))
        dzy = (sum * 100) / totalCellSize
    }
}
